### PR TITLE
Pass socket FDs to all ExecXYZ= commands in socket unit but ExecStartPre=

### PR DESCRIPTION
**Pull request by jsitnicki**
This PR comes from the discussion we had in issue #6714.

The situation today is that the `ExecXYZ=` commands from a socket unit can't access the socket FDs that the unit creates.

This means that to perform some one-off action on the socket FD, we need to resort to other solutions. Like setting up an extra `oneshot` service unit, in addition to the service unit actually servicing the socket. This is a bit verbose and has some caveats, like `NonBlocking=` setting must match in both service definitions.

What kind of one-off action we may want to perform on a socket FD?

The main motivation is to be able to store a socket reference into a BPF map. Like `BPF_MAP_TYPE_SOCKHASH` map, which can be then used to dispatch connections to sockets, or `BPF_MAP_TYPE_REUSEPORT_SOCKARRAY` map which we can use for load-balancing within a reuseport group (collection of sockets which share a local address). We might also just want to take the FD and send it to some other process, one which perhaps has access to BPF maps.

I have put together an example to illustrate the use case at:

https://github.com/jsitnicki/systemd-playground/blob/socket-pass-fds-to-exec/example.mixin/etc/systemd/system/daytime.socket

To enable the use case, we pass socket file descriptors to `ExecXYZ=` commands from the socket unit. As an exception, `ExecStartPre=` command does not inherit any FDs because it runs before the sockets are created.

This translates to what is not supported by this feature. We cannot tweak any socket options that need to be set before a socket gets bound to an address. All `ExecXYZ=` commands but `ExecStartPre=` run after the sockets have been created and bound.

The new behavior is guarded by a new `PassFileDescriptorsToExec=` socket option.  To avoid breaking any existing setups it defaults to `off`.

New functionality is covered by an an integration test (`TEST-85-SOCKET-PASS-FDS`). It was simpler to write a short shell script and a couple socket units, but it can be converted to a C-based unit test under `src/test` if that is preferable.

### CODE REVIEW

File: tools/dbus_ignorelist
Comment (YHNdnzj) [inline tools/dbus_ignorelist]: No. This is for methods/properties that don't need explicit version info, i.e. added before v183. Please add proper version info for new properties.

**Review by YHNdnzj — COMMENTED** [review]
I think the first two commits should be squashed? If you are guarding the new behavior behind a flag, do it immediately rather than in another commit?

File: test/TEST-85-SOCKET-PASS-FDS/test.sh
Comment (YHNdnzj) [inline test/TEST-85-SOCKET-PASS-FDS/test.sh]: No need to add a whole new test for this. Please just add a subtest in TEST-07-PID1.

Comment (jsitnicki) [reply test/TEST-85-SOCKET-PASS-FDS/test.sh]: Will do. Thanks for the hint.

File: tools/dbus_ignorelist
Comment (jsitnicki) [reply tools/dbus_ignorelist]: Oh, okay. Let me address that.

File: src/core/socket.c
Comment (YHNdnzj) [inline src/core/socket.c]: Please move these into the if block.

File: src/core/socket.c
Comment (YHNdnzj) [inline src/core/socket.c:1927]: Please add a comment here explaining that we don't pass fds to ExecStartPre= intentionally

File: src/core/socket.c
Comment (YHNdnzj) [inline src/core/socket.c]: Hmm, so EXEC_PASS_FDS is currently only used in service_spawn_internal, i.e. has no effect on execute.c/exec-invoke.c. But it feels more correct to set it here.

**Comment by jsitnicki** [comment]
Thanks for feedback, @YHNdnzj. Will address it and circle back with a v2.
(I'm preparing for KubeCon next week, so it will take me a few days.)

**Comment by jsitnicki** [comment]
Changes in v2:
- introduce new behavior and new option in one commit
- update History section in `org.freedesktop.systemd1` man page
- move new locals in `socket_spawn()` into the relevant block
- set `EXEC_PASS_FDS` flag `socket_spawn()` when passing FDs
- add explanatory comment `socket_spawn()` on `ExecStartPre=`
- incorporate integration test into existing `TEST-07-PID1` test

File: man/systemd.socket.xml
Comment (YHNdnzj) [inline man/systemd.socket.xml]: We don't usually say "XYZ" like this. Maybe:

> The passed file descriptors can be accessed with `sd_listen_fds`, as if the commands were invoked from the associated service units.

You get the idea ;-)

**Review by YHNdnzj — APPROVED** [review]
LGTM. I'd like to wait for @poettering's final review though.

File: man/systemd.socket.xml
Comment (jsitnicki) [reply man/systemd.socket.xml]: That does sound better. Updated. Thanks.

**Comment by jsitnicki** [comment]
Changes in v3:
- reworded `PassFileDescriptorsToExec` man page section as per @YHNdnzj suggestion

**Comment by yuwata** [comment]
Hm, why we need to introduce the new knob? In what situation the new behavior breaks existing setups?
I think we should unconditionally pass fds, and announce the change in NEWS.

**Comment by yuwata** [comment]
> FWIW, the knob has been suggested by @poettering in a [related discussion](https://github.com/systemd/systemd/issues/6714#issuecomment-326346440).
> 
> Happy to remove it, though. Makes things simpler.

Thanks. The following his comment https://github.com/systemd/systemd/issues/6714#issuecomment-326515170 answers my question.

OK. LGTM.

**Review by yuwata — APPROVED** [review]

**Comment by yuwata** [comment]
(BTW, why several .service units, e.g. systemd-bootctl@.service, explicitly set the env var `Environment=LISTEN_FDNAMES=varlink` ?)

**Comment by poettering** [comment]
> (BTW, why several .service units, e.g. [systemd-bootctl@.service](mailto:systemd-bootctl@.service), explicitly set the env var `Environment=LISTEN_FDNAMES=varlink` ?)

See varlink_invocation(). It's a routine that detects automatically if a tool is invoked in a varlink context or not. This test is done by looking for an fd being passed that carries the "varlink" fd name.

**Pull request merged**
